### PR TITLE
refactor: add checkDmain()

### PR DESCRIPTION
**Pull request by WalterBright**
Makes the code flow a little nicer.

Also picked up a couple -> to . conversions that had been overlooked.

### CODE REVIEW

File: src/func.d [hunk 1]
```diff
@@ -4018,6 +3991,36 @@ extern (C++) class FuncDeclaration : Declaration
...
+        if (nparams == 1)
+        {
+            auto fparam0 = Parameter.getNth(tf.parameters, 0);
+            auto t = fparam0.type.toBasetype();
+            if (t.ty != Tarray ||
+                t.nextOf().ty != Tarray ||
+                t.nextOf().nextOf().ty != Tchar ||
+                fparam0.storageClass & (STCout | STCref | STClazy))
+            {
+                argerr = true;
```
Comment (andralex) [inline src/func.d]: why not issue the argument error straight here?

File: src/func.d [hunk 2]
```diff
@@ -4018,6 +3991,36 @@ extern (C++) class FuncDeclaration : Declaration
...
+            }
+        }
+
+        if (!tf.nextOf())
+            error("must return int or void");
+        else if (tf.nextOf().ty != Tint32 && tf.nextOf().ty != Tvoid)
+            error("must return int or void, not %s", tf.nextOf().toChars());
+        else if (tf.varargs || nparams >= 2 || argerr)
+            error("parameters must be main() or main(string[] args)");
+    }
```
Comment (andralex) [inline src/func.d]: this code is in desperate need of trailing parentheses elimination

**Review by andralex — APPROVED** [review]

File: src/func.d (hunk 1 above)
Comment (WalterBright) [inline src/func.d]: So it's behavior remained the same as the previous code w.r.t. what errors are printed for each sort of error.

File: src/func.d (hunk 2 above)
Comment (WalterBright) [inline src/func.d]: ?

**Comment by WalterBright** [comment]
Auto-merge toggled on

File: src/func.d (hunk 2 above)
Comment (andralex) [inline src/func.d]: e.g. `tf.nextOf().nextOf().toChars()` -> `tf.nextOf.nextOf.toChars`

Comment (WalterBright) [inline src/func.d]: It's C++, not D!

Comment (andralex) [inline src/func.d]: I stand destroyed.

Comment (WalterBright) [inline src/func.d]: Another one of those "crap, this is C++, not D!" moments I sometimes have :-)

**Pull request merged**